go/logic/migrator.go

```diff
@@ -500,6 +520,22 @@ func (this *Migrator) cutOver() (err error) {
 	this.migrationContext.MarkPointOfInterest()
 	this.migrationContext.Log.Debugf("checking for cut-over postpone: complete")
 
+	this.migrationContext.Log.Infof("Waiting for heartbeat lag to be low enough to proceed")
+	this.sleepWhileTrue(
+		func() (bool, error) {
+			heartbeatLag := this.migrationContext.TimeSinceLastHeartbeatOnChangelog()
+			maxLagMillisecondsThrottleThreshold := atomic.LoadInt64(&this.migrationContext.MaxLagMillisecondsThrottleThreshold)
+			if heartbeatLag > time.Duration(maxLagMillisecondsThrottleThreshold)*time.Millisecond {
+				this.migrationContext.Log.Debugf("current HeartbeatLag (%.2fs) is too high, it needs to be less than --max-lag-millis (%.2fs) to continue", heartbeatLag.Seconds(), (time.Duration(maxLagMillisecondsThrottleThreshold) * time.Millisecond).Seconds())
+				return true, nil
```

Comment:
Running two distinct `sleepWhileTrue` loops poses a problematic scenario where possibly the conditions tested in the first loop can re-appear while sleeping on the 2nd loop. I can imagine a user running a `echo "postpone" | socat - "/tmp/gh-ost.sock"` while sleeping on the newly introduced heartbeat lag loop.

My suggestion is to consolidate the two tests inside one loop, and I think it's simple: remove this new `sleepWhileTrue` and move the above five rows to be the fist condition tested in the original loop. Overall, there will be one `sleepWhileTrue` loop that'll look like this:

  
```
	this.migrationContext.MarkPointOfInterest()
	this.migrationContext.Log.Debugf("checking for cut-over postpone")
	this.sleepWhileTrue(
		func() (bool, error) {
                        heartbeatLag := this.migrationContext.TimeSinceLastHeartbeatOnChangelog()
                        maxLagMillisecondsThrottleThreshold := atomic.LoadInt64(&this.migrationContext.MaxLagMillisecondsThrottleThreshold)
                        if heartbeatLag > time.Duration(maxLagMillisecondsThrottleThreshold)*time.Millisecond {
                          this.migrationContext.Log.Debugf("current HeartbeatLag (%.2fs) is too high, it needs to be less than --max-lag-millis (%.2fs) to continue", heartbeatLag.Seconds(), (time.Duration(maxLagMillisecondsThrottleThreshold) * time.Millisecond).Seconds())
                          return true, nil
                        }
			if this.migrationContext.PostponeCutOverFlagFile == "" {
				return false, nil
			}
			if atomic.LoadInt64(&this.migrationContext.UserCommandedUnpostponeFlag) > 0 {
				atomic.StoreInt64(&this.migrationContext.UserCommandedUnpostponeFlag, 0)
				return false, nil
			}
			if base.FileExists(this.migrationContext.PostponeCutOverFlagFile) {
				// Postpone file defined and exists!
				if atomic.LoadInt64(&this.migrationContext.IsPostponingCutOver) == 0 {
					if err := this.hooksExecutor.onBeginPostponed(); err != nil {
						return true, err
					}
				}
				atomic.StoreInt64(&this.migrationContext.IsPostponingCutOver, 1)
				return true, nil
			}
			return false, nil
		},
	)
	atomic.StoreInt64(&this.migrationContext.IsPostponingCutOver, 0)
	this.migrationContext.MarkPointOfInterest()
	this.migrationContext.Log.Debugf("checking for cut-over postpone: complete")
```

what do you think?

Reply:
I'd possibly change:
```
if heartbeatLag > time.Duration(maxLagMillisecondsThrottleThreshold)*time.Millisecond {
```
 to 
```
if heartbeatLag > time.Duration(base.CutOverLockTimeoutSeconds)*time.Second {
```
It depends on whether a user's setting of `maxLagMillisecondsThrottleThreshold` is higher or lower than `maxLagMillisecondsThrottleThreshold`. In my previous use case, lag threshold was `1sec` and cut-over timeout was `3sec`, and it made more sense to initiate cut-over if lag was within bounds of `3sec`.

Perhaps even the higher of the two values?

Reply:
> Running two distinct sleepWhileTrue loops poses a problematic scenario where possibly the conditions tested in the first loop can re-appear while sleeping on the 2nd loop

I completely missed this, you are absolutely correct. I have pushed a commit `Consolidate the two sleepWhileTrue loops` to address this.

Reply:
> In my previous use case, lag threshold was 1sec and cut-over timeout was 3sec, and it made more sense to initiate cut-over if lag was within bounds of 3sec.
> 
> Perhaps even the higher of the two values?

This makes a lot of sense to me, I think this is a good improvement. I have pushed a commit `HeartbeatLag must be < than --max-lag-millis and --cut-over-lock-time` to address this.